### PR TITLE
feat(mimo): add Phase 5 checkpointing and evaluation infrastructure

### PR DESCRIPTION
**Pull request by aroshanghias-nvd**
## Summary

- Add `MimoOptimizerWrapper` for unified optimizer state checkpointing across heterogeneous modules
- Add `slice_batch_for_mimo` utility in `dp_utils.py` for DP-aware batch slicing
- Integrate `save_checkpoint` with `MimoOptimizerWrapper` in `train_mimo.py`
- Add evaluation support via `evaluate_and_print_results` with multimodule communicator
- Add async checkpoint finalization with `maybe_finalize_async_save`
- Add unit tests for `MimoOptimizerWrapper` and `slice_batch_for_mimo` (11 tests passing)

## Files Changed

| File | Description |
|------|-------------|
| `mimo_checkpointing.py` | New - `MimoOptimizerWrapper` for unified optimizer checkpointing |
| `dp_utils.py` | Added `slice_batch_for_mimo` for DP-aware batch slicing |
| `train_mimo.py` | Added checkpointing, evaluation, async finalization |
| `pretrain_mimo.py` | Minor cleanup |
| `test_mimo_checkpointing.py` | New - Unit tests |

## Test plan

- [x] Unit tests pass: `pytest tests/unit_tests/training/test_mimo_checkpointing.py -v` (11 passed)
- [ ] Integration testing with MIMO training (pending MIMO optimizer implementation)

## Summary by CodeRabbit

* **New Features**
  * Added comprehensive MIMO (Multi-Input Multi-Output) model support for training heterogeneous multi-module architectures
  * Introduced multi-modal data loading pipeline supporting concurrent processing of images, audio, and text inputs
  * Enabled heterogeneous distributed training with per-module parallelism configuration and data-parallel aware batching

* **Tests**
  * Added extensive unit test coverage for MIMO data loading, model providers, training utilities, and configuration validation

### CODE REVIEW

File: src/megatron/bridge/data/loaders.py [hunk 1]
```diff
@@ -175,6 +175,27 @@ def build_train_valid_test_data_loaders(
...
+        train_samples, valid_samples, test_samples = get_train_valid_test_num_samples(cfg)
+        return build_mimo_data_loaders(
+            cfg=cfg,
+            train_state=train_state,
+            mimo_provider=cfg.dataset,
+            train_samples=train_samples,
+            valid_samples=valid_samples,
+            test_samples=test_samples,
+        )
+
```
[the remark this answers is not in the document]
Comment (aroshanghias-nvd) [reply src/megatron/bridge/data/loaders.py:211]: Fixed. Added `all_reduce(MAX)` to sync `train_state` flags across all ranks before returning from the MIMO path.

File: src/megatron/bridge/data/mimo/hf_provider.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    encoder_seq_lengths: Dict[str, int] = field(default_factory=dict)
+    modality_columns: Dict[str, str] = field(default_factory=dict)
+    text_column: str = "text"
+    train_split: str = "train"
+    valid_split: str = "validation"
+    test_split: str = "test"
+    
+    # Cached processors and tokenizer (loaded once)
+    _processors: Optional[Dict[str, Any]] = field(default=None, repr=False)
+    _tokenizer: Optional[Any] = field(default=None, repr=False)
```
[the remark this answers is not in the document]
Comment (aroshanghias-nvd) [reply src/megatron/bridge/data/mimo/hf_provider.py:80]: Fixed. Added `trust_remote_code: Optional[bool] = None` field to the dataclass.

File: src/megatron/bridge/data/mimo/mock_provider.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+            trust_remote_code=is_safe_repo(
+                trust_remote_code=self.trust_remote_code,
+                hf_path=self.tokenizer_path,
+            ),
+        )
+        
+        if tokenizer.pad_token is None:
+            tokenizer.pad_token = tokenizer.eos_token
+        
+        object.__setattr__(self, "_tokenizer", tokenizer)
```
[the remark this answers is not in the document]
Comment (aroshanghias-nvd) [reply src/megatron/bridge/data/mimo/mock_provider.py:140]: Fixed. Added validation at the start of `_load_tokenizer()` that raises a clear ValueError if `tokenizer_path` is empty.

File: src/megatron/bridge/models/mimo/mimo_ddp.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
+"""DDP wrapping utilities for MIMO models.
+
+Called from the training layer after MimoModelProvider.provide().
+
+Note: This module only supports DDP wrapping. FSDP is not yet implemented.
+"""
```
[the remark this answers is not in the document]
Comment (aroshanghias-nvd) [reply src/megatron/bridge/models/mimo/mimo_ddp.py:7]: Fixed. Added NVIDIA copyright header.

File: src/megatron/bridge/models/mimo/mimo_provider.py [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+        if final_pre_wrap_hook:
+            result = final_pre_wrap_hook(model_list)
+            if result is not None:
+                model_list = result
+        
+        # Move to device
+        if not use_cpu_initialization and not init_model_with_meta_device:
+            for m in model_list:
+                m.cuda(torch.cuda.current_device())
+
```
[the remark this answers is not in the document]
Comment (aroshanghias-nvd) [reply src/megatron/bridge/models/mimo/mimo_provider.py:431]: Fixed. Added resolution logic to use provider defaults when `use_cpu_initialization` or `init_model_with_meta_device` are None.

File: src/megatron/bridge/training/eval.py [hunk 6]
```diff
@@ -14,18 +14,23 @@
...
 from megatron.core.pipeline_parallel.p2p_communication import P2PCommunicator
 from megatron.core.pipeline_parallel.utils import is_pp_last_stage
+from megatron.core.process_groups_config import ProcessGroupCollection
 from megatron.core.rerun_state_machine import RerunDataIterator, RerunMode, get_rerun_state_machine
 from megatron.core.transformer import MegatronModule
 from megatron.core.utils import get_model_config
 
+# TODO: Pending PR 3129
+from megatron.core.pipeline_parallel.multimodule_communicator import MultiModulePipelineCommunicator
+from megatron.core.process_groups_config import MultiModuleProcessGroupCollection
```
[the remark this answers is not in the document]
Comment (aroshanghias-nvd) [reply src/megatron/bridge/training/eval.py]: Fixed. Made imports conditional with try/except and added `_MULTIMODULE_AVAILABLE` flag for runtime checks.

File: src/megatron/bridge/training/train_mimo.py [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+    
+    # Prepare forward step function with GlobalState injection
+    wrapped_forward_step_func = prepare_forward_step_func(forward_step_func, global_state)
+    
+    # Build module-to-grid mapping for gradient operations
+    module_to_grid_tuple = get_module_to_grid_tuple(model, mimo_infra)
+    
+    # Build pg_collection for schedule
+    multimodule_pg_collection = build_pg_collection_for_schedule(mimo_infra)
+
```
[the remark this answers is not in the document]
Comment (aroshanghias-nvd) [reply src/megatron/bridge/training/train_mimo.py:216]: Fixed. Added guard that raises RuntimeError if `multimodule_pg_collection` is a list (fallback not supported for MIMO training).

File: src/megatron/bridge/training/train_mimo.py [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+    total_loss_dict = {}
+    history_wct = []
+    report_memory_flag = True
+    
+    # TODO: Revisit when MIMO optimizer is implemented.
+    # MimoOptimizerWrapper aggregates per-module optimizer states for checkpointing.
+    # Currently uses first scheduler only - may need MimoSchedulerWrapper for consistency.
+    optimizer_wrapper = MimoOptimizerWrapper(optimizers)
+    first_scheduler = next(iter(schedulers.values()), None) if schedulers else None
+
```
[the remark this answers is not in the document]
Comment (aroshanghias-nvd) [reply src/megatron/bridge/training/train_mimo.py]: All MIMO modules share the same LR schedule and are stepped synchronously, so their scheduler states are identical. Saving the first scheduler's state is sufficient. A MimoSchedulerWrapper would only be needed if we add per-module LR schedules in the future.

**Comment by aroshanghias-nvd** [comment]
Consolidated into PR #2137 (feat(mimo): MIMO training, checkpointing, and evaluation infrastructure).

Phase 5 content (checkpointing, evaluation, parallelism constraints, e2e tests) has been merged into the Phase 4 training PR for streamlined review. All changes from this PR are preserved in #2137.
